### ISSUE TITLE
VAES support

### ISSUE THREAD

**Issue by nazar-pc** [RustCrypto/block-ciphers#372]
Vectorized AES can process more than one block at a time, greatly improving throughput, but it doesn't appear to be used by `aes` crate yet, which is unfortunate.

**Comment by nazar-pc** [comment]
Absolutely, I am aware of that. Already using `*_block` and there is a difference indeed.

Those methods are the reason I looked inside wondering if it is also VAES-capable on top of instruction parallelism.

**Comment by newpavlov** [comment]
Unfortunately, currently I do not have access to a machine with AVX-512 so can not work on it myself.

If someone will work on this, it also could be worthwhile to also add support of VPCLMULQDQ to the `polyval`crate.

**Comment by nazar-pc** [comment]
Shouldn't be difficult to get server VM with AVX512 support, I can help with that if you'd like.

**Comment by newpavlov** [comment]
@tarcieri 
I think it should be a relatively straightforward implementation? You would need to play with number of 512-bit blocks processed in parallel, because VAES instructions may have a different latency/throughput ratio and compiler can use 32 registers instead of 16. I think an optimal number will be bigger than the 8 blocks used in the AES-NI backend. IIUC VAES does not have instructions for key expansion, so you can adapt the existing AES-NI-based code.

The only annoying part should be plumbing in the autodetection module, but we can leave it for later, since VAES support will be gated either way (the relevant intrinsics are Nightly-only right now).

>polyval is/was written in a way that LLVM will already use VPCLMULQDQ when the target supports it:

I thought about using parallelism (VPCLMULQDQ can process four 128-bit blocks at once). But I forgot that GHASH/POLYVAL is inherently sequential, so we can not utilize it while processing one text...

**Comment by tarcieri** [comment]
@newpavlov I opened a separate issue for VPCLMULQDQ here, I think that should be (potentially) fairly easy: https://github.com/RustCrypto/universal-hashes/issues/184

POLYVAL/GHASH can be broken down into a parallelizable portion and a sequential portion... there's an accumulation of the output that is inherently sequential, but multiplication of the inputs can be performed in parallel.

**Comment by nazar-pc** [comment]
Funnily enough, a few years later I'm back to this issue after seeing XMM registers instead of ZMM when feeding 8 blocks at a time to `encrypt_blocks`/`decrypt_blocks`.

If there is a need to access the hardware, I can create a VM on my workstation that is AVX512-capable (AMD Zen 4).

At least on Intel CPUs the throughput is 1 CPI for 512-bit/4 blocks encoding/decoding, which is very appealing to me and should be 2x faster than doing 1 block at a time with CPI 0.5.

**Comment by silvanshade** [comment]
> There was a PR to add VAES support, but the author decided to close it when it was almost ready to merge, for whatever reason: [#396](https://github.com/RustCrypto/block-ciphers/pull/396)

The reason was it sat there for 6 months with no proper review despite making all the changes requested.

You also never actually said you would even merge it or even laid out conditions for what would be required to get it to a merge ready state so from my point of view "almost ready" wasn't a thing.

You and @newpavlov had way more than enough opportunity to respond and follow through if you actually valued the contribution.

**Comment by silvanshade** [comment]
Here is another opportunity to merge it: https://github.com/RustCrypto/block-ciphers/pull/482
